File: framework/languages/dotnet/e2e/SpotService/Client/Scenarios/SmE3Scenario.cs
using SpotService.Client;
using SpotService.Shared;
using Zlink.HttpClient;

namespace SpotService.Client.Scenarios;

internal static class SmE3Scenario
{
    public static async Task RunAsync(ZLinkHttpClient playA)
    {
        var spotRid = $"spot-sm-e3-{Guid.NewGuid():N}";
        var created = (await playA.Post("/spot/create")
            .Body(new CreateSpotReq(spotRid))
            .SubmitAsync<CreateSpotReply>()).Body;
        ScenarioAssert.That(created.SpotRid == spotRid && created.NodeRid == "play-a", "SM-E3 idle spot was not created on play-a.");
        var idle = (await playA.Post("/spot/idle-close/start")
            .Body(new SpotIdleCloseReq(spotRid, "sm-e3-idle", 50))
            .SubmitAsync<SpotIdleCloseReply>()).Body;
        ScenarioAssert.That(idle.Closed, "SM-E3 idle close did not close the spot.");
        var closedSpotRequest = (await playA.Post("/spot/missing-target/request")
            .Body(new SpotMissingTargetReq(spotRid))
            .SubmitAsync<SpotMissingTargetReply>()).Body;
        ScenarioAssert.That(closedSpotRequest.Failed, "SM-E3 closed spot request did not fail.");
        await EvidenceWait.ForAllAsync(
            playA,
            [
                $"timer-idle-close|rid=play-a|spot={spotRid}|name=sm-e3-idle|closed=True",
                $"spot-closing|rid=play-a|spot={spotRid}",
            ],
            "SM-E3 evidence mismatch.");
        Console.WriteLine("operation SpotService.sm-e3 passed");
    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-e3-scenario.ts
// SM-E3: idle timer 기반 명시적 close 시나리오를 검증한다.
import type {
  CreateSpotRes,
  CreateSpotReq,
  EvidenceWaitReq,
  SpotIdleCloseRes,
  SpotIdleCloseReq,
  SpotMissingTargetRes,
  SpotMissingTargetReq
} from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../Support/http-client';
import { ensure } from '../Support/scenario-assert';

export async function runSmE3(options: ClientOptions): Promise<void> {
  const spotRid = `spot-sm-e3-${Date.now()}`;
  const created = await postJson<CreateSpotRes>(options.playAUrl, '/spot/create', {
    spotRid
  } satisfies CreateSpotReq);
  ensure(created.spotRid === spotRid && created.nodeRid === 'play-a', 'SM-E3 idle spot was not created on play-a.');

  const idle = await postJson<SpotIdleCloseRes>(options.playAUrl, '/spot/idle-close/start', {
    spotRid,
    name: 'sm-e3-idle',
    periodMs: 50
  } satisfies SpotIdleCloseReq);
  ensure(idle.closed, 'SM-E3 idle close did not close the spot.');

  const closedSpotRequest = await postJson<SpotMissingTargetRes>(options.playAUrl, '/spot/missing-target/request', {
    spotRid
  } satisfies SpotMissingTargetReq);
  ensure(closedSpotRequest.failed, 'SM-E3 closed spot request did not fail.');

  const expectedEvidence = [
    `timer-idle-close|rid=play-a|spot=${spotRid}|name=sm-e3-idle|closed=True`,
    `spot-closing|rid=play-a|spot=${spotRid}`
  ];
  const evidence = await postJson<string[]>(options.playAUrl, '/evidence/wait', {
    containsAll: expectedEvidence,
    timeoutMilliseconds: 10000
  } satisfies EvidenceWaitReq);
  ensure(
    expectedEvidence.every((expected) => evidence.some((line) => line.includes(expected))),
    'SM-E3 evidence mismatch.'
  );

  console.log('scenario SM-E3 passed');
}
